### DISCUSSION TITLE
Is it possible to have multi-character surrounds? (Plus a different, unrelated question.)

### DISCUSSION THREAD

**Discussion by 6601618158** [kylechui/nvim-surround#232] [Q&A]
I hope I'm using the correct terminology here, but what I'm aiming at is basically whether it is possible to replicate something like this: 

```
Old text Command New text ~
|a = b|  Sm(     \left(a = b\right) 
```

This is supposed to show that the visual selection "a = b" is surrounded by "\left(" and "\right)", meaning I can use "m(" to refer to the pair "\left(" and "\right)" (ideally for adding, changing, and removing). 

My goal would be to basically replicate the (La)TeX specific mappings from vim-sandwich, as shown here: https://github.com/machakann/vim-sandwich/wiki/Filetype-tex. 

----

_Sorry for the long text that follows:_

I also have another, rather unrelated question. Ever since I switched from Emacs to (Neo)vim for basically all of my text editing, I've been looking for a replacement for smartparens (https://github.com/Fuco1/smartparens). 

In addition to adding, deleting, and changing ‘surrounds’, like what this plugin can do, there are also features like ‘slurping’, ‘barfing’, ‘splitting’, and ‘joining’ (among others, but those are the things I miss most). What those things mean is basically the following (I'll try to illustrate with simple examples; old → new, and ^ marking the cursor position): 

Slurping (forward), meaning ‘baz’ is ‘pulled in’:
```
[foo bar] baz → [foo bar baz]
        ^               ^
```

Barfing (forward), meaning ‘baz’ is ‘spit out’ again: 
```
[foo bar baz] → [foo bar] baz
        ^               ^
```

(This is also explained here: https://ebzzry.com/en/emacs-pairs/#slurpandbarf). 

Splitting: 
```
[foo bar baz] → [foo bar] [baz]
        ^                ^
```

Joining: 
```
[foo bar] [baz] → [foo bar baz]
         ^                ^
```

Are operations like these at all in the scope of this plugin? 

‘Slurping’ and ‘barfing’, although more cumbersome, can basically be replicated by using something like nvim-gomove (https://github.com/booperlv/nvim-gomove), visually selecting the `]`, and then just moving it around, but I haven't found a way to ‘split’ or ‘join’ without manually typing everything out, of course. But then you might get stuff like
```
[foo bar baz]  →  [foo bar][]baz]
        ^     i][          ^
```
(the part under the → means I'm entering insert mode), because an auto-close/auto-pair plugin interferes, which is kind of frustrating. 

My ‘idea’ for an implementation would basically involve adding a ‘join’ and a ‘split’ operator (like the ‘add’, ‘delete’, and ‘change’ operators), give it the ‘surround’ you want to use (i.e. manually), and then remove/insert the corresponding pair, like this (just for this example: yJ = ‘join’ [you join], yS = ‘split’ [you split]): 

```
Old text             Command     New text
"hello"* "world"     yJ"         "hello world"
"hello* world"       yS"         "hello"* "world"
```
Ideally, you could also do something like this:

```
Old text              Command     New text
(a + b) *+ (c + d)    yJ(         (a + b *+ c + d)
(a + b *+ c + d)      yS(         (a + b) *+ (c + d)
```
and maybe this (the function/operator would default to using the ‘surround’ on the left if they're different but part of the same alias)
```
Old text             Command     New text
"hello"* 'world'     yJq         "hello world"
```
----

As far as I understand, vim-sexp (https://github.com/guns/vim-sexp) can do ‘slurping’ and ‘barfing’ (there it's called ‘capturing’ and ‘emitting’), but the plugin seems to be kind of old (and in viml), would duplicate a lot of functionality I don't need because plugins like this one can already do it, and joining and splitting doesn't seem to be supported, at all. 

On the other hand, there seem to be various parinfer-plugins (https://shaunlebron.github.io/parinfer/), but parinfer AFAIK only works for Lisp (dialects) _and_ does these kinds of things automatically, _somehow_ (‘slurping’ and ‘barfing’, at least; no idea whether joining and splitting works).

**Comment by kylechui** [comment]
Hi there!
* There is currently no plan for supporting multi-character surrounds like you want, although that is not out of the cards if enough people want it. My main concerns are:
  * Additional complexity/overhead in the codebase
  * Confusion among users, as then the surrounds would need to form a prefix-free encoding in order to be properly utilized

Here are a few suggestions/alternatives:
* Assuming you are using VimTeX, there is the built-in `tsd` (toggle surrounding delimiters) that changes parens/brackets/curly braces to their respective `\left`/`\right` forms
* `nvim-surround` supports a `buffer_setup` call, which allows you to set your configuration per buffer, so something like setting `(` to perform `\left(` and `\right)` while perhaps setting `b` to do `(` and `)` is possible *only for LaTeX files* (if you decide to do this, I can help you set that up)

As for the slurping and barfing, I actually haven't heard of that before (nor do I know of any existing plugins that do that); perhaps I will want to make it happen (though probably not as a part of this plugin, as it seems to involve moving things around as opposed to simply changing what the delimiters are), as it is a problem that I have run into a few times :thinking: Could probably use Tree-sitter or something to automatically decide what to "slurp up" as well, or perhaps use Vim motions to do that instead.

**Comment by gaoqiangks** [comment]
Nvim-surrounds already supports multi-character surrounds,  only that you have to implement it youself.  The bottom is my config, and I hope it can help you.   

Let me explain a bit.   For example, if I want to place the current line into a theorem environment, then I will press `ySSe` in normal mode, and then nvim asks me to input the targeting environment name. So I input `theorem`, and press `<CR>`. Now the current will be put into a theorem environment.  

Another example. We have a sentence like the following.
`If a+b=b+a for any a and b, then we say + is commutative`.

Now I want to bold the word `commutative`.  Select this word in visual mode, press `S`, and then press `c`, then nvim asks you to input the name of a command. So we input `textbf`, and hit `<CR>`. Now the sentence will becomes 

`If a+b=b+a for any a and b, then we say + is \textbf{commutative}`.

Alternatively, you can also use `ysawc`, and then input `textbf`.




```

if vim.fn.has_key(vim.api.nvim_get_var('plugs'), 'nvim-surround')~=0 then
	require("nvim-surround").setup({
		move_cursor=false,
		--aliases=
		--{
			--["b"]="",
		--},
		--nvim-surround的操作是可重复的. 比如ysiw", 将一个单词用"引用后, 在下一个单词可以直接按.来重复
		surrounds = 
		{
			[")"] = {
				add = function()
					return { { "( " }, { " )" } }
				end,
			},
			["("] = {
				add = function()
					return { { "(" }, { ")" } }
				end,
			},
			["}"] = {
				add = function()
					return { { "{ " }, { " }" } }
				end,
			},
			["{"] = {
				add = function()
					return { { "{" }, { "}" } }
				end,
			},
			["["] = {
				add = function()
					return { { "[" }, { "]" } }
				end,
			},
			["]"] = {
				add = function()
					return { { "[ " }, { " ]" } }
				end,
			},
			["<"] = {
				add = function()
					return { { "<" }, { ">" } }
				end,
			},
			[">"] = {
				add = { "< " }, { " >" },
			},
			["E"] = {
				add = function()
					return {{ "\\begin{equation}" }, {"\\end{equation}"}}
				end,
			},
			["T"] = {
				add = function()
					return {{ "\\begin{theorem}" }, {"\\end{theorem}"}}
				end,
			},
			["e"] = {
				add = function()
					--ySSe 将当前行放在环境中. 比如*eq, 将会放在equation*环境中. eq会放在equation环境中.
					local config = require("nvim-surround.config")
					local env_input = config.get_input("surround输入环境名称:")
					local env_map = {
						["eq"]="equation",
						["th"]="theorem",
						["pr"]="proposition",
						["le"]="lemma",
						["re"]="remark",
						["co"]="corollary",
						["ex"]="exercise",
						["em"]="example",
						["de"]="definition",
						["sp"]="split",
						["qu"]="question",
					}
					local env = {
						["equation"]=1,
						["theorem"]=1,
						["proposition"]=1,
						["lemma"]=1,
						["remark"]=1,
						["corollary"]=1,
						["exercise"]=1,
						["example"]=1,
						["definition"]=1,
						["split"]=1,
						["problem"]=1,
						["question"]=1,
					}
					local final_env=nil
					local no_n=0
					if env_input==nil then
						return
					end
					local first_letter=string.sub(env_input,1,1)
					if first_letter=="*" then
						no_n=1
						env_input=string.sub(env_input,2,#env_input)
					end
					--local s=env[env_input]
					--if s then
						--final_env=env_input
					--else
						--local first_two_letters=string.sub(env_input,1,2)
						--if first_two_letters then
							--final_env=env_map[first_two_letters]
						--else
					for k, v in pairs(env) do
						--string.find反回两个值, 第一个值是匹配开始的地方, 第二个值是匹配结束的地方
						ibegin, iend=string.find(k, env_input)
						if ibegin==1 then
							final_env=k
							break
						end
					end
						--end
					--end
					if final_env then
						if no_n==1 then
							final_env=final_env.."*"
						end
					else
						final_env=env_input
					end
					return { {"\\begin{"..final_env.."}"}, {"\\end{"..final_env.."}" }}
				end,
			},
			--ysiwi 将当前单词放在textit环境中
			["i"] = {
				add = { "\\textit{", "}" },
				find = "\\textit%b{}",
				delete = "^(\\textit{)().-(})()$",
			},
			["1"] = {
				add = { "\\textbf{", "}" },
				find = "\\textbf%b{}",
				delete = "^(\\textbf{)().-(})()$",
			},
			["b"] = {
				add = { "\\mathbb{", "}" },
				find = "\\mathbb%b{}",
				delete = "^(\\mathbb{)().-(})()$",
			},
			--ysiwc 将当前单词放在命令中
			--vimtex提供了dsc  csc dic等删除, 修改命令等操作, 但是没有提供ys**
			["c"] = {
				add = function()
					local config = require("nvim-surround.config")
					local cmd_input = config.get_input("surround输入命令名称:")
					if cmd_input==nil then
						return
					end
					return {{"\\"..cmd_input.."{"}, {"}"}}
				end,
				find = "\\mathcal%b{}",
				delete = "^(\\mathcal{)().-(})()$",
			},
		}
	})
end


````
